File: 3.py


class  Solution:
    def lengthOfLongestSubstring(self, s :str) -> int:
        
        l = 0
        r = 0
        ans = 0
        hm = {}
        
        for r in range(0,len(s)):
            hm[s[r]] = hm.get(s[r] , 0) + 1
            

            while hm[s[r]] > 1:
                hm[s[l]] -= 1
                l += 1
                
            ans = max(ans, r - l + 1)
                
                
            
        
        return ans
                
                
            
                
                
            
            
            
        
s = Solution()

ans = s.lengthOfLongestSubstring("abcabcbb")

print(ans)

File: 3.cpp
#include <iostream>
#include <string>
#include <unordered_map>

using namespace std;




class Solution {
public:
    int lengthOfLongestSubstring(string s) {
        int currMaxLen{};
        int l{};
        int r{};
        unordered_map<char, int> m;

        if(s.size() == 0) return 0;


        while(s.size() > r){
            if(m.count(s[r])){
                l = max(l, m[s[r]] +1); 
            }    
            m[s[r]] = r;

            currMaxLen = max(currMaxLen, r - l + 1);
            r++;
        }

        return currMaxLen;

    }
};


int main(){

    Solution s;

    string test  = "abcabcbb";

    int ans = s.lengthOfLongestSubstring(test);

    cout<<ans<<endl;

    return 0;
}